[AIRFLOW-2110] Enhance Http Hook to use a header in passed in the "extra" argument and add tenacity retry

albertocalderari opened the pull request:
Make sure you have checked _all_ steps below.

## JIRA

- [x]   My PR addresses the following Airflow JIRA issues and references them in the PR title. For example, "[AIRFLOW-XXX] My Airflow PR"
[AIRFLOW-2110] Enhance Http Hook to use a header in passed in the "extra" argument and add tenacity retry"
https://issues.apache.org/jira/browse/AIRFLOW-2122

## Description
- [x]  Add possibility to add a json header in the "extra" field in the connection:
{"Authorization": "Bearer Here1sMyT0k3N"}
Also add tenacity retry so the operator won't fall over in case of a connection issue.
Here are some details about my PR, including screenshots of any UI changes:
In the original hook we could not skip raise for status, pass headers from the "extra" field in the airflow UI and would immediately fail on connection error. Now in case of connection error the hook will retry 7 times with exponential backoff, the "raise for status" method can be skipped and can pass a JSON header in the extra field in the UI.

## Tests

- [x]  My PR adds the following unit tests OR does not need testing for this extremely good reason:
test_raise_for_status_with_200
test_get_request_do_not_raise_for_status_if_check_response_is_false
test_hook_contains_header_from_extra_field
test_hook_uses_provided_header
test_hook_has_no_header_from_extra
test_hooks_header_from_extra_is_overridden
test_post_request
test_post_request_with_error_code
test_post_request_do_not_raise_for_status_if_check_response_is_false
test_retry_on_conn_error
test_header_from_extra_and_run_method_are_merged

## Commits 

- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "How to write a good git commit message":

Subject is separated from body by a blank line
Subject is limited to 50 characters
Subject does not end with a period
Subject uses the imperative mood ("add", not "adding")
Body wraps at 72 characters
Body explains "what" and "why", not "how"

- [x]  Passes git diff upstream/master -u -- "*.py" | flake8 --diff

Commit:
[AIRFLOW-2110] Use "extra" connection parameter for the header if specified

Commit:
[AIRFLOW-2110] added tenacity retry and separate check response from run_and_check

Commit:
[AIRFLOW-2110] extracted method to get airflow connection from main get conn

Extracted method to get airflow connection from main get conn to
simplify testing

Commit:
[AIRFLOW-2110] Added unittest and fixed compatibility issues

[AIRFLOW-2110] added post and get tests

[AIRFLOW-2110] Added connection error retry test and used proper mocking

[AIRFLOW-2110] Fix backward compatibility issue

[AIRFLOW-2110] Fix Flake8 issue

[AIRFLOW-2110] Fix Flake8 issues on tests file

[AIRFLOW-2110] Fix mock issue and use self.log instead of logging

[AIRFLOW-2110] Fix breaking change on http_sensor

[AIRFLOW-2110]  Fix Py3 Compatibility issue

ashb commented on tests/hooks/test_http_hook.py:
```diff
@@ -0,0 +1,213 @@
+        with mock.patch(
+            'airflow.hooks.base_hook.BaseHook.get_connection',
+            side_effect=get_airflow_connection_noheaders
+        ):
```
Mocking connections seems like a more complex than just creating/updating an actuall connection object. Let's just do that. (we can re-use the "http_default" connection here)

albertocalderari replied:
@ashb I did this because the default "http_default" connection points to google.com and did not want to create a dependency on the input value of the connection. But if this looks really far too complex for you I'll change it

ashb replied:
Since we are mocking the requests here anyway, I think it’s cleaner to just use that. (Tell me if you try it and the code looks worse)

albertocalderari replied:
@ashb I do agree that the code would look a lot cleaner without the mock, which BTW I noticed is also absolutely useless -.- . I'll remove it.

ashb commented on tests/hooks/test_http_hook.py:
```diff
@@ -0,0 +1,213 @@
+    return conn
+
+
+class TestHttpHook(unittest.TestCase):
```
A new test file is good, but could you pull the tests out of tests/core.py (`class HttpHookTest`) and put them in here too please?

albertocalderari replied:
@ashb did not even notice there was a test suit for the http hook, ill merge them together

ashb commented on tests/hooks/test_http_hook.py:
```diff
@@ -0,0 +1,213 @@
+            self.get_hook.run.retry.stop.max_attempt_number + 1,
+            mocked_session.call_count
+        )
+
```
Could you also add a test that combines headers from the connection with headers passed in to the `hook.run()` call

albertocalderari replied:
@ashb Added this => test_header_from_extra_and_run_method_are_merged

ashb commented on scripts/ci/requirements.txt:
```diff
@@ -89,6 +89,7 @@ Sphinx-PyPI-upload
 sphinx_rtd_theme
 sqlalchemy>=1.1.15, <1.2.0
 statsd
+tenacity==4.8.0
```
This file has been removed now, but more importantly this only makes it get installed under CI -- this is needed at run-time and should go in setup.py somewhere appropriate.

ashb replied:
(Oh this file doesn't exist anymore on master, but this still needs to be listed in a non-test only dependency)

albertocalderari replied:
@ashb Sure, must have missed it before

ashb commented on airflow/hooks/http_hook.py:
```diff
@@ -48,11 +57,18 @@ def get_conn(self, headers):
         if conn.login:
             session.auth = (conn.login, conn.password)
+        if conn.extra:
+            session.headers.update(json.loads(conn.extra))
```
There's an `extra_dejson` property that we should use here

ashb reviewed (changes requested):
Some sort of advanced retry sounds useful, but I'm not sure if it is right in this form:

1) It has hard-coded values that mean this is going to do the wrong thing for some people.
2) There are probably other places where this might be wanted. Is it possible to just use this at the call site instead -- something like `retry(hook.run)(args...)` ?

Commit:
[AIRFLOW-2110] Added tenacity to setup.py

ashb commented on airflow/hooks/http_hook.py:
```diff
@@ -85,34 +101,36 @@ def run(self, endpoint, data=None, headers=None, extra_options=None):
+            return response
+
+        except requests.exceptions.ConnectionError as ex:
+            logging.error(str(ex) + 'add retry logic here')
```
Oh missed this first time -this should be `self.log.error` ... and we probably want something else here?

Commit:
[AIRFLOW-2110] Moved tests from core to http hook and added new test case

Added test case test_header_from_extra_and_run_method_are_merged

Commit:
[AIRFLOW-2110] made error message more understandable

Commit:
[AIRFLOW-2110] Code review changes

[AIRFLOW-2110] Fix accidentally broken test

[AIRFLOW-2110] Remove useless mock

[AIRFLOW-2110] Remove unused import

[AIRFLOW-2110] Use extra_dejson property and change Mock

[AIRFLOW-2110] remove unused import

[AIRFLOW-2110] remove unused import

Commit:
[AIRFLOW-2110] Abstract advance retrying into base hook

Fokko commented on tests/hooks/test_http_hook.py:
```diff
@@ -0,0 +1,260 @@
+from airflow.exceptions import AirflowException
+
+try:
+    from airflow.hooks.http_hook import HttpHook
```
This should always be importable

albertocalderari replied:
I copied these from core - happy to change the behaviour though

Fokko replied:
Please

Fokko commented on tests/hooks/test_http_hook.py:
```diff
@@ -0,0 +1,260 @@
+    )
+
+
+@unittest.skipIf(HttpHook is None,
```
The test should always run

albertocalderari replied:
I copied these from core - happy to change the behaviour though

Fokko replied:
Maybe we should also clean up core in a different PR

Fokko commented on airflow/hooks/base_hook.py:
```diff
@@ -96,3 +96,18 @@ def get_pandas_df(self, sql):
     def run(self, sql):
         raise NotImplementedError()
+
+    def run_with_advanced_retry(self, _retry_args, *args, **kwargs):
```
I don't think we should add this in the `base_hook`, this should be just in the `http_hook`

albertocalderari replied:
happy to move it back to http_hook

Fokko reviewed (changes requested):
@albertocalderari nice PR. I have a few comments, please take a look.

Fokko commented on airflow/hooks/http_hook.py:
```diff
@@ -85,34 +93,36 @@ def run(self, endpoint, data=None, headers=None, extra_options=None):
+            return response
+
+        except requests.exceptions.ConnectionError as ex:
+            self.log.error(str(ex) + ' Tenacity will retry to execute the operation')
```
Could  you make this `warn`, since it is not fatal.

Fokko commented on setup.py:
```diff
@@ -243,6 +243,7 @@ def do_setup():
             'sqlalchemy>=1.1.15, <1.2.0',
             'sqlalchemy-utc>=0.9.0',
             'tabulate>=0.7.5, <0.8.0',
+            'tenacity == 4.8.0',
```
Nit, remove spaces :-)

Commit:
[AIRFLOW-2110] Code review

Commit:
[AIRFLOW-2110] Added HttpHook docs and example
